Adds the power flow control console

Xhuis opened the pull request:
:cl: Xhuis
add: Added the power flow control console, which allows remote manipulation of most APCs on the z-level. You can find them in the Chief Engineer's office on all maps.
/:cl:

The power flow control console, or APC control for short, lets the user remotely access APCs. The console can access all APCs on the z-level with a few exceptions:
* High-security areas (classified as AI-monitored) cannot be accessed.
* If the APC's AI control wire is cut, it cannot be accessed.
* Broken APCs cannot be accessed.
* APCs with their panel open cannot be accessed.

To use the console, you need to log in using a valid Engineering ID. When used, the console displays all connected APCs, along with their area, charge, and status. You can interface with any APC by clicking on its name; when you do this, you will be able to modify the APC like any silicon can.

When an APC is remotely accessed, its _interface unlocks to all users_ and it says an audible notice to anyone within hearing range. The interface is locked again when the remote access terminates. Interface lock cannot be changed from remote access.

To help make it more usable, the console's interface has different filters. You can sort by name, charge, and accessibility. The interface updates every tick until you leave.

This console is intended to let engineers control more finely where power flows and to some degree allow some wiggle room and strategy. If the power is out because xenos are around and people need to get through to the weapons the Armory has, you can shut down nonessential area APCs to let Security stay powered. This also has nefarious uses, of course. That is partially offset by the warning and interface unlock, and it will become abundantly clear if someone is tampering with every APC remotely.

Boards can be built in R&D with programming 4, engineering 4, and power storage 5. (I think I still need to do the recipe, though.) Additionally, every major station has been outfitted with a console somewhere in the engine room, usually replacing the redundant atmospheric alert console.

Some pictures:
_[An unfiltered interface](https://puu.sh/uEyA8/c298f59621.png)_
_[Remotely accessing an APC](https://puu.sh/uEyIe/c54828cea7.png)_
_[A one-minute demo showing off the console's general use](https://puu.sh/uFMxN/fa54e2d98e.gif)_

_This is the first part of some Engineering changes I'm making to flesh out the role a little more and give them something to do besides fix breaches and build forts. All feedback is highly encouraged._

Commit:
Adds the power flow control console

kevinz000 commented:
@Xhuis Just one suggestion, can you make a wire or just make the AI control wire be able to be cut to disable this?
Having a counter would be great.
Or, actually, if this works on wire connection, doesn't even matter, as if you really don't want people screwing with you that badly you can just cut the wire and get a cell. Either would be nice, thanks.

Xhuis commented:
As stated very early in the OP:
> If the APC's AI control wire is cut, it cannot be accessed.

Supermichael777 commented:
make it need a id inserted

Xhuis commented:
As stated in the OP:
>Right now, anyone can use the console. I would like to restrict access to it

RemieRichards commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,109 @@
+
+/obj/machinery/computer/apc_control/process()
+	apcs = list() //Clear the list every tick
+	for(var/obj/machinery/power/apc/APC in apcs_list)
```
istypeless

Xhuis replied:
Knew I forgot something.

RemieRichards commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,109 @@
+	popup.set_content(dat)
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.open()
+	return
```
trailing

RemieRichards commented on code/modules/power/apc.dm:
```diff
@@ -681,7 +681,7 @@
 		"totalLoad" = lastused_total,
 		"coverLocked" = coverlocked,
-		"siliconUser" = user.has_unlimited_silicon_privilege,
+		"siliconUser" = user.has_unlimited_silicon_privilege || locate(/obj/machinery/computer/apc_control) in range(1, user),
```
make this validate that user is apc_control's user.

Xhuis replied:
Through some testing this isn't very cooperative to pull off. This is snowflakey but I think it's the best solution I have. For context, here's what I tried:

```
var/obj/machinery/computer/apc_control/A = locate() in range(1, user)
if(A)
  data["siliconUser"] = A.owner == user
```

That was below the rest of the assignment.

RemieRichards replied:
you could just make it ```|| using_power_console(user)``````
/proc/using_power_console(mob/user)
...
```

Xhuis replied:
Ah, that could work. I don't like global procs, but hey! This thing might have some more uses added eventually, so I guess so.

Xhuis commented:
Do you watch for my PRs like a hawk or is this just routine? Are you stalking me?

RemieRichards commented:
who?

Xhuis commented:
You, silly!

RemieRichards commented:
no I'm a maintainer, one who's not turned their emails off, so I get instantly notified of new PRs, this one just caught my interest

Xhuis commented:
The first step to admitting you like someone is to admit that you've been following them~

RemieRichards commented:
I like you but not in that way.

octareenroon91 commented:
I suggest mapping the console somewhere a little bit deeper in Engineering than the nuke-ops entrance.

PKPenguin321 commented:
I'd like to see this function by powernet, so you can also beat it by snipping your wires and going off the grid

KorPhaeron commented:
Might be better as a Cheif Engineer thing but this is a good PR

FantasticFwoosh commented:
@PKPenguin321 so you mean that it would just read only the APC's its connected to? What about if they are emagged open or AI hacked, wouldn't that be a massive red flag if its notified?

(you know so people don't abuse APC's that aren't meant to be seen)

coiax commented:
@Xhuis I like you, b-baka

Commit:
Minor cleanup

kevinz000 commented:
@Xhuis lol i can't read sry kek

MrPerson commented:
The only issue with this is that it exposes further issues with engineering and power.

Hatfish commented:
Would you consider adding this to modular computers as well, or would that be out of the scope of this PR?

Xhuis commented:
I don't know how modular computers work, so probably not.
@MrPerson I'm going to do some work on those later, too!

GunHog commented:
How does this interact with emagged APCs? No change?
Can silicons use the console?
Does it preserve the locked/unlocked state of APCs unlocked via APC or hacking?
Do APCs respond to commands while the wires are exposed or the cover is open?

coiax commented:
@Xhuis maybe you could make it a program that runs on a console (possibly prevent tablets for balance) reasons, rather than a dedicated computer?

I have no idea how our program code works here.

GunHog commented:
@Shadowlight213 is needed for such a thing, @coiax.

coiax commented:
Then we must say his name three times while throwing salt over our shoulders.

@Shadowlight213

lzimann commented:
Make sure the TGUI closes when you move away from the console(an early gif you posted on coderbus didn't seem to have that happening).

MrStonedOne commented:
If we are going to do remote access, we need to have more in game logging and tracking as a downside.

How about having all operations logged, and have it so any power monitor console can access those logs.

Xhuis commented:
that's a good idea! I'll do that. also reminds me that I need to do admin logging for this.

Shadowlight213 commented:
I was summoned.
Also looking at the code, it could easily work as a program.
Especially since unlike the power monitor it doesn't use a direct power net connection.

Making a program isn't that hard. Look at the program file for the power net console for reference I guess.

Also, there's a flag for limiting programs to certain hardware, so balancing would be easy. The power monitor actually had to be limited to consoles because of the physical link.

Screemonster commented:
This is something that a lot of codebases have been missing for a longass time. 👍

KorPhaeron commented:
It probably doesnt need to be a program or even constructable. Some assistant controlling the stations power from space or maint wouldnt be much fun

Xhuis commented:
Right now you can make boards from R&D but it has semi-high research requirements. I want to provide a way to get it back if the console is sabotaged or destroyed, and needing Engineering access is an effective way of stopping abuse.

Xhuis commented:
Also, @GunHog:
1. In the next commit they will not be able to interface with emagged APCs.
2. Untested but I don't see why not.
3 No.
4. Untested but I believe they do.

Xhuis commented:
Here's a preview of the logging.
![](https://puu.sh/uFCSo/b0b82e23ce.png)
Timestamps will be changed, obviously.

Commit:
Adds some new stuff

Commit:
Tweaks sounds

Xhuis commented:
* Added a logging feature. All activity in the console will be kept in an internal log. Emagging the computer will cause the operator to not be logged, and also allow for someone to clear the logs.
* APCs with open panels will no longer respond to interface attempts.
* Emagged APCs do not appear on the list.
* The interface no longer updates automatically, as it was causing the scroll bar to reset and this is incredibly annoying.
* Added sounds to the console as well as APCs when they're being interfaced with.

KorPhaeron commented:
@tgstation/commit-access 

I think remote control of the stations power is something that should belong to CE access not engineering access does anyone agree or disagree

Xhuis commented:
If maintainers are uncertain then I can move it into the CE's office. I was apprehensive about it, though, considering how much the CE gets already.

Actually, that reminds me. Let me add access requirements.

KorPhaeron commented:
All the CE gets right now is a pile of toys, this would be a practical way for him to have authority

GunHog commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,164 @@
+		if(check_apc(APC))
+			apcs[APC.name] = APC
+	if(operator)
+		if(!operator.Adjacent(src))
```
If you do decide to allow silicon control, please remember that you need to check for 'operator.has_unlimited_silicon_privilege' or 'issilicon(operator)` (Only one, not both) such that they are not locked out by the Adjacent check.

If you do NOT want silicons to control the console, then override `attack_ai()` with a simple `return` such that you do not have oddities like putting an AI directly next to the console, or a borg simply rolling up to it.

Xhuis commented:
Yeah @KorPhaeron, but they're _shiny_ toys.
@GunHog I'll probably end up disallowing AI control because they already have APC access essentially everywhere.

Commit:
Added access requirements, admin logging, AI restriction

Xhuis commented:
* The console now requires an Engineering-access ID to use.
* AIs can no longer use the console.
* Remote access now messages admins and adds to the game log.

It seems like the prevailing opinion here is CE only, so I'll move it to their office.

Commit:
Part 1: Remove dem consoles

Commit:
Put dem consoles back

Xhuis commented:
The consoles have been moved into the CE's office. Access requirements remain the same in case he goes MIA (which, being a head, may be likely.) The logging also means that anyone trying to sneakily use the console is ousted as soon as someone checks them.

On OmegaStation, there doesn't seem to _be_ a CE office, so I kept it in the SMES room.

Xhuis commented:
Unless anyone has anything to add, I think this is good!

Commit:
Small patch

Xhuis commented:
* You can now bypass access requirements by using an emag. This gives an indication in the form of the authentication ID.
* Logins/logouts are now logged.
* APCs can no longer be modified if the console is depowered.

Commit:
Some work I can't remember

Xhuis commented:
* You can now undo emagging through a button at the bottom of the console.

Also updated the OP with a 1-minute demo of the console in its current state.

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+	desc = "Used to remotely control the flow of power to different parts of the station."
+	icon_screen = "solar"
+	icon_keyboard = "power_key"
+	req_access = list(access_engine)
```
shouldn't this be a string or something

Xhuis replied:
Nope. That's how access works. Sadly.

Iamgoofball replied:
this is better than a string, though?

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+	var/checking_logs = 0
+	var/list/logs
+	var/authenticated = 0
+	var/auth_id = "\[NULL\]"
```
this is a weird syntax

Xhuis replied:
It's used for display. It appears as [NULL] in the console.

Iamgoofball replied:
yeah gotta escape those characters

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+		to_chat(usr, "<span class='robot notice'>\icon[I] Logging functionality restored from backup data.</span>")
+		emagged = FALSE
+		LAZYADD(logs, "<b>-=- Logging restored to full functionality at this point -=-</b>")
+	if(href_list["access_apc"])
```
you need to recheck all assumptions in the topic call, distance fo player, player status, if the apc is valid to intefere with etc

cannot rely on byond to only provide valid topic calls.

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+		active_apc = APC
+	if(href_list["name_filter"])
+		playsound(src, 'sound/machines/terminal_prompt.ogg', 50, 0)
+		var/new_filter = stripped_input(usr, "What name are you looking for?", name) as null|text
```
these are sleeps, so you need to check the item isn't qdeleted and all the assumptions are still valid (emag valid, player still alive and not deleted etc)

putting all the checks in canUseTopic and then calling that in all the plcaes would be good

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+		filters["Name"] = new_filter
+	if(href_list["above_filter"])
+		playsound(src, 'sound/machines/terminal_prompt.ogg', 50, 0)
+		var/new_filter = input(usr, "Enter a percentage from 1-100 to sort by (greater than).", name) as null|num
```
sleeps as well

optimumtact reviewed (changes requested):
Also why did you recompile tgui if you're not using it?

Xhuis commented:
The tgui files are left over from when I tried to put this in tgui, then saw my mistake. They shouldn't have an effect, but I can cut them out if you'd like.

optimumtact commented:
I'd prefer it, just to avoid any unexpected issues

Shadowlight213 commented:
@KorPhaeron @xhuis technically making a program downloadable is just a var on it.

As for making it rnd built. I should look into making rnd constructable disks with programs on them.

GunHog commented:
@Shadowlight213 Please teach me how to make programs or create a detailed Wiki guide that can be understood by a 12 year old!

Xhuis commented:
I mean I'm 5 so

Xhuis commented:
My local git isn't picking up a difference between the tgui files, so I think we're good. Other changes on the way.

Commit:
Sanity checkin'

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+				log_activity("logged in")
+		if(!authenticated) //Check for emags
+			var/obj/item/weapon/card/emag/E = usr.get_active_held_item()
+			if(E && istype(E))
```
you should check adjacency for this

Xhuis replied:
It can only happen if the user activates it while standing next to the console, so I don't think it's needed

optimumtact replied:
you cannot trust the topic call to do that, someone can forge the call and hold the emag item in their hand

Iamgoofball replied:
^^^^

Commit:
Maybe now!

optimumtact commented on code/game/machinery/computer/apc_control.dm:
```diff
@@ -0,0 +1,208 @@
+	if(href_list["access_apc"])
+		playsound(src, "terminal_type", 50, 0)
+		var/obj/machinery/power/apc/APC = locate(href_list["access_apc"]) in apcs_list
+		if(!APC || APC.aidisabled || APC.panel_open)
```
what if it's qdeleted? check that too

Iamgoofball replied:
we should really make a default proc for checking basic hrefsploit problems

Xhuis replied:
I can probably whip something up later today.

Commit:
Sanity checks pass 3295912

Penguaro commented:
This is great! I've been looking into a similar concept for Air Alarms and this gives me a better idea of how to proceed. Great work!

Commit:
Sanity checks
